Sync/Synchronizers/Bookmarks/BookmarksSynchronizer.swift

```diff
@@ -120,18 +120,66 @@ open class BufferingBookmarksSynchronizer: TimestampedSingleCollectionSynchroniz
             run = doMirror >>== { result in
-                // Just validate to report statistics.
-                if case .completed = result {
-                    log.debug("Validating completed buffer download.")
-                    return buffer.validate().bind { validationResult in
-                        if let invalidError = validationResult.failureValue as? BufferInvalidError {
-                            self.statsSession.validationStats = self.validationStatsFrom(error: invalidError)
+                guard case .completed = result else {
+                    return deferMaybe(result)
+                }
+                return storage.getLocalBookmarksAdditions() >>== { newBookmarks -> Success in
+                    guard newBookmarks.count > 0 else {
+                        return succeed()
+                    }
+                    // This assumes that "Mobile Bookmarks" exists in the buffer, which will be false if no computer ever connected to that sync account
+                    return buffer.getBufferItemAndChildrenWithGUID(BookmarkRoots.MobileFolderGUID) >>== { curMobileRoot in
+                        let newChildren = newBookmarks.values.map { $0.guid }
+                        var allChildren = curMobileRoot.children ?? []
+                        // We do not use a set in order to keep the children's order
+                        for child in newChildren {
+                            if !allChildren.contains(child) {
+                                allChildren.append(child)
+                            }
+                        }
+                        let mobileRoot = curMobileRoot.copyWithChildren(children: allChildren)
+
+                        var bookmarks = newBookmarks
+                        bookmarks[mobileRoot.guid] = mobileRoot
+
+                        let records = Array(bookmarks.values.map { bkm -> Record<BookmarkBasePayload> in
+                            let payload = bkm.asPayload()
+                            let mappedGUID = payload["id"].string ?? bkm.guid
+                            return Record<BookmarkBasePayload>(id: mappedGUID, payload: payload)
+                        })
+
+                        var uploaded = [GUID]()
+                        let onUpload: (POSTResult, Timestamp?) -> DeferredTimestamp = { result, lastModified in
+                            uploaded.append(contentsOf: result.success.map { BookmarkRoots.translateIncomingRootGUID($0) })
+
+                            let modified = lastModified ?? 0
+                            for guid in result.success {
+                                bookmarks[guid]?.serverModified = modified
+                            }
+                            return deferMaybe(modified)
+                        }
+
+                        return self.uploadRecords(records, lastTimestamp: info.modified(self.collection) ?? 0, storageClient: bookmarksClient, onUpload: onUpload) >>== { lastModified -> Success in
```

Comment:
I'm not sure that `info.modified` will be correct here — we fetch `i/c`, then we do some work, then we download bookmarks, and now we upload them. Someone could have uploaded data between the `i/c` fetch and the bookmarks fetch, and Sync is totally OK with that.

You should use the last download timestamp of the bookmarks synchronizer instead — that is, you want to make sure that this upload happens immediately after the download.